### PR TITLE
[camera] android-rework part 2: Android auto focus feature

### PR DESCRIPTION
**Pull request by mvanbeusekom**
Adds the `AutoFocusFeature` containing the implementation to handle auto focus via the Android Camera2 API.

This is the second PR in a series of pull-requests which will gradually introduce changes from PR #3651, making it easier to review the code (as discussed with @stuartmorgan).

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
  - Since this will be a series of pull-requests I will update the version in the last PR of the series.
- [ ] I updated CHANGELOG.md to add a description of the change.
  - Since this will be a series of pull-requests I will update the version in the last PR of the series.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by acoutts** [comment]
@googlebot I consent.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import android.hardware.camera2.CameraCharacteristics;
+import android.hardware.camera2.CaptureRequest;
+import io.flutter.plugins.camera.CameraProperties;
+import io.flutter.plugins.camera.features.CameraFeature;
+
+public class AutoFocusFeature extends CameraFeature<FocusMode> {
+  private FocusMode currentSetting = FocusMode.auto;
+
+  // When we switch recording modes we re-create this feature with
+  // the appropriate setting here.
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java:16]: I have no idea who/what the "we"s in this sentence refer to. It doesn't even feel like they are referring to the same thing; setting a private field would be done internally to the class, but recreating the feature would be done from outside the class. Please use clear subjects in the comment instead.

"we" is almost always problematic in (non-TODO) comments, as this ambiguity is extremely common; I advise against using it in comments.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  private final boolean recordingVideo;
+
+  public AutoFocusFeature(CameraProperties cameraProperties, boolean recordingVideo) {
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java:19]: Please add Javadoc

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @Override
+  public boolean checkIsSupported() {
+    int[] modes = cameraProperties.getControlAutoFocusAvailableModes();
+
+    // Check if fixed focal length lens. If LENS_INFO_MINIMUM_FOCUS_DISTANCE=0, then this is fixed.
+    // Can be null on some devices.
+    final Float minFocus = cameraProperties.getLensInfoMinimumFocusDistance();
+    // final Float maxFocus = cameraCharacteristics.get(CameraCharacteristics.LENS_INFO_HYPERFOCAL_DISTANCE);
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java:46]: Why is there commented out code?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    // Value can be null on some devices:
+    // https://developer.android.com/reference/android/hardware/camera2/CameraCharacteristics#LENS_INFO_MINIMUM_FOCUS_DISTANCE
+    boolean isFixedLength;
+    if (minFocus == null) {
+      isFixedLength = true;
+    } else {
+      isFixedLength = minFocus == 0;
+    }
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java:55]: Consider a ternary so the assignment and declaration are the same statement: `boolean isFixedLength = minFocus == null ? true : (minFocus == 0);`

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Override
+  public void updateBuilder(CaptureRequest.Builder requestBuilder) {
+    if (!checkIsSupported()) {
+      return;
+    }
+
+    switch (currentSetting) {
+      case locked:
+        /** If we're locking AF we should do a one-time focus, then set the AF to idle */
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java:70]: s/If we're locking AF we should do/When locking AF do/

Although it's not entirely clear to me what this comment is describing; is the line that is being commented actually setting auto-focus to idle?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
A few small things, but looks good overall. (I only reviewed the new files; I'm assuming there were no changes to the files that are part of "part 1")

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java (hunk 2 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/autofocus/AutoFocusFeature.java:27]: done

**Review by stuartmorgan — APPROVED** [review]
LGTM, sorry for the delay.

**Pull request merged**